### PR TITLE
Initial delivery

### PR DESCRIPTION
**Pull request by lazypower**
Adds the scaffolding for GitHub and google oid connect providers.
Additionally, adds some quality of life improvements for users via
converting config keys to jinja2 friendly _'s.

Adds cleanup logic during the stop hook, health checks, and reactions to
config-changed events.

### CODE REVIEW

**Comment by lazypower** [comment]
Looking for initial feedback here. its pretty basic, the scaffolding is starting to come together but there's still a few miles to go in terms of the integration. Nothing useable here outside of the [example app](https://github.com/coreos/dex/blob/master/Documentation/getting-started.md) which is `lazy-dex.example-app`

If you attempt to run it now all you'll get is the token inspector as an output. K8s integration is still TODO

File: README.md [hunk 1]
```diff
@@ -1,34 +1,18 @@
...
-
-and so on. If you're providing a web service or something that the end user
-needs to go to, tell them here, especially if you're deploying a service that
-might listen to a non-default port.
-
-You can then browse to http://ip-address to configure the service.
+```
+juju deploy cs:~containers/easyrsa
+juju deploy cs:~containers/dex
+juju add-relation dex easyrsa
```
Comment (lazypower) [inline README.md:13]: This bit isn't finished either but I was kind of guiding my thoughts this way that we'll need to securely wrap this service behind TLS to ensure it's not susceptible to MITM.

File: config.yaml [hunk 2]
```diff
@@ -1,14 +1,41 @@
 options:
-  string-option:
+  github-client:
```
Comment (ktsakalozos) [inline config.yaml:2]: Do we have intend to have so tight integration with github and google? How would we extend this? Add config variables here and update the template?

Comment (lazypower) [reply config.yaml:2]: They were the initial targets. I think for now the idea is to integrate with the low hanging fruit and progress towards adding in the other integration paths like adding SAML/LDAP

**Comment by marcoceppi** [comment]
+1

**Pull request merged**